File: 26-sum/prog/td-game/src/game/enemy.cpp
#include "enemy.h"

#include <algorithm>
#include <random>

#include "game.h"
#include "path.h"
#include "status_effect.h"

Enemy::Enemy(
  std::string const& type,
  Rect bounds,
  float speed,
  int health,
  int tower_damage,
  float tower_damage_cooldown
)
    : type_(type),
      bounds_(bounds),
      speed_(speed),
      health_(health),
      max_health_(health),
      tower_damage_(tower_damage),
      tower_damage_cooldown_(tower_damage_cooldown) { }
Enemy::~Enemy() = default;

std::string const& Enemy::type() const { return type_; }
Rect Enemy::bounds() const { return bounds_; }
Vec2 Enemy::position() const { return bounds_.center; }
int Enemy::health() const { return health_; }
int Enemy::max_health() const { return max_health_; }
void Enemy::set_health(int health) { health_ = health; }
void Enemy::increase_health(int amount) { health_ += amount; }
void Enemy::decrease_health(int amount) { health_ -= mitigate(amount); }
bool Enemy::is_destroyed() const { return health_ <= 0; }
int Enemy::tower_damage() const { return tower_damage_; }
float Enemy::tower_damage_cooldown() const { return tower_damage_cooldown_; }
int Enemy::score_value() const { return score_; }
void Enemy::set_score(int s) { score_ = s; }

Path const* Enemy::path() const { return path_; }
void Enemy::set_path(Path const* path) { path_ = path; }
float Enemy::path_distance() const { return path_distance_; }
void Enemy::set_path_distance(float d) { path_distance_ = d; }

bool Enemy::portal_ready() const { return portal_cooldown_.is_finished(); }
void Enemy::start_portal_cooldown(float duration) {
  portal_cooldown_ = CountdownTimer(duration, true);
}

float Enemy::slow_resist_factor() const { return 1.0f; }
float Enemy::splash_damage_factor() const { return 1.0f; }
int Enemy::mitigate(int damage) const { return std::max(0, damage); }
void Enemy::on_death(Game&) { }

void Enemy::apply_status_effect(std::unique_ptr<StatusEffect> effect) {
  effect->on_apply(*this);
  status_effects_.push_back(std::move(effect));
}

void Enemy::sync_position() {
  if (path_) {
    bounds_.center = path_->position_at(path_distance_);
  }
}

Vec2 Enemy::predicted_position(float time_ahead) const {
  if (path_) {
    return path_->position_at(path_distance_ + speed_ * time_ahead);
  }
  return bounds_.center;
}

Enemy::StatusHint Enemy::status_hint() const {
  StatusHint hint;

  for (auto const& effect : status_effects_) {
    if (effect->expired()) continue;
    switch (effect->type()) {
    case StatusEffect::Type::Slow:
      hint.slow = true;
      break;
    case StatusEffect::Type::Poison:
      hint.poison = true;
      break;
    case StatusEffect::Type::Regen:
      hint.regen = true;
      break;
    }
  }

  return hint;
}

void Enemy::update(Game& game, float dt) {
  bool rooted = false;
  float slow_mult = 1.0f;

  // Compute the effective speed multiplier from all active status effects and terrain.
  for (auto const& effect : status_effects_) {
    if (!effect->expired()) {
      rooted = rooted || effect->roots();
      slow_mult *= effect->speed_multiplier();
    }
  }

  Tile const* tile = game.map().tile_at(position());
  float terrain_speed_factor = tile ? tile->enemy_speed_factor() : 1.0f;

  // Update velocity factor and position based on terrain and status effects.
  float velocity_factor;
  if (rooted) {
    velocity_factor = 0.0f;
  } else {
    velocity_factor = terrain_speed_factor * std::min(1.0f, slow_mult * slow_resist_factor());
  }

  if (path_) {
    path_distance_ += speed_ * dt * velocity_factor;
    bounds_.center = path_->position_at(path_distance_);
  }

  // Update all status effects and remove expired ones.
  for (auto& effect : status_effects_) {
    effect->update(*this, dt);
  }

  status_effects_.erase(
    std::remove_if(
      status_effects_.begin(),
      status_effects_.end(),
      [](auto const& effect) { return effect->expired(); }
    ),
    status_effects_.end()
  );

  // Update the portal cooldown timer.
  portal_cooldown_.update(dt);
}

ResistantEnemy::ResistantEnemy(
  std::string const& type,
  Rect bounds,
  float speed,
  int health,
  int tower_damage,
  float tower_damage_cooldown,
  float slow_resist,
  float splash_resist
)
    : Enemy(type, bounds, speed, health, tower_damage, tower_damage_cooldown),
      slow_resist_(slow_resist),
      splash_resist_(splash_resist) { }

float ResistantEnemy::slow_resist_factor() const { return slow_resist_; }
float ResistantEnemy::splash_damage_factor() const { return splash_resist_; }

SplitterEnemy::SplitterEnemy(
  std::string const& type,
  Rect bounds,
  float speed,
  int health,
  int tower_damage,
  float tower_damage_cooldown,
  ChildSpec const& child_spec
)
    : Enemy(type, bounds, speed, health, tower_damage, tower_damage_cooldown),
      child_spec_(child_spec) { }

void SplitterEnemy::on_death(Game& game) {
  for (int i = 0; i < child_spec_.count; ++i) {
    auto child = std::make_shared<Enemy>(
      "normal",
      Rect(bounds_.center, child_spec_.width, child_spec_.height),
      child_spec_.speed,
      child_spec_.health,
      child_spec_.tower_damage,
      child_spec_.tower_damage_cooldown
    );
    child->set_path(path_);
    // Perturb the child's path distance so siblings spread along the path
    // instead of stacking on the exact same spot.
    float dist = path_distance_;
    if (child_spec_.perturbation > 0.0f) {
      float offset =
        std::uniform_real_distribution<float>(-child_spec_.perturbation, child_spec_.perturbation)(
          game.rng()
        );
      dist = std::max(0.0f, dist + offset);
      if (path_ && dist >= path_->total_length())
        dist = path_->total_length() - 0.01f;  // don't reach the exit
    }
    child->set_path_distance(dist);
    child->sync_position();
    child->set_score(child_spec_.score);
    game.spawn_enemy(std::move(child));
  }
}

BossEnemy::BossEnemy(
  std::string const& type,
  Rect bounds,
  float speed,
  int health,
  int tower_damage,
  float tower_damage_cooldown,
  int shield,
  float regen_amount,
  float regen_duration,
  float regen_interval
)
    : Enemy(type, bounds, speed, health, tower_damage, tower_damage_cooldown),
      shield_(shield),
      regen_amount_(regen_amount),
      regen_duration_(regen_duration),
      regen_interval_(regen_interval),
      regen_timer_(regen_interval, true) { }

int BossEnemy::mitigate(int damage) const { return std::max(0, damage - shield_); }

void BossEnemy::update(Game& game, float dt) {
  Enemy::update(game, dt);

  if (regen_interval_ > 0.0f) {
    regen_timer_.update(dt);
    if (regen_timer_.is_finished()) {
      apply_status_effect(std::make_unique<RegenerationEffect>(regen_amount_, regen_duration_));
      regen_timer_.reset();
    }
  }
}

File: 26-sum/prog/td-game/src/rust/src/enemy.rs
use rand::RngExt;
use serde::{Deserialize, Serialize};

use crate::{
    geometry::{Rect, Vec2},
    map::Map,
    path::Path,
    status_effect::{StatusEffect, StatusEffectKind},
    timer::CountdownTimer,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChildSpec {
    pub width: f32,
    pub height: f32,
    pub speed: f32,
    pub max_health: i32,
    pub score: i32,
    pub tower_damage: i32,
    pub tower_damage_interval: f32,
    pub count: i32,
    /// Max random offset (in path-distance units) from parent.
    pub perturbation: f32,
}

#[derive(Debug, Clone)]
pub enum EnemyKind {
    Normal,
    Fast,
    Armored,
    /// Resistant to slow and to splash damage.
    Resistant {
        slow_resist: f32,
        splash_resist: f32,
    },
    /// Spawns smaller enemies on death.
    Splitter {
        child_spec: ChildSpec,
    },
    /// Significantly stronger enemy with a damage-absorbing shield and regeneration.
    Boss {
        /// Reduces incoming damage by `shield`.
        shield: i32,
        /// Amount of health to regenerate per second.
        regen_amount: i32,
        /// Duration of the regeneration effect.
        regen_duration: f32,
        /// Apply regen effect every `interval` seconds, 0.0 = no regeneration.
        regen_interval: f32,
        regen_cooldown_timer: CountdownTimer,
    },
}

impl EnemyKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Normal => "normal",
            Self::Fast => "fast",
            Self::Armored => "armored",
            Self::Resistant { .. } => "resistant",
            Self::Splitter { .. } => "splitter",
            Self::Boss { .. } => "boss",
        }
    }
}

/// An enemy follows the map's path toward the exit and can attack towers and be
/// attacked by bullets.
#[derive(Debug, Clone)]
pub struct Enemy {
    pub bounds: Rect,
    pub speed: f32,
    pub health: i32,
    pub score: i32,
    pub max_health: i32,
    pub tower_damage: i32,
    pub tower_damage_interval: f32,
    pub path_index: usize,
    pub path_distance: f32,
    pub portal_cooldown_timer: CountdownTimer,
    pub status_effects: Vec<StatusEffect>,
    pub kind: EnemyKind,
}

slotmap::new_key_type! { pub struct EnemyId; }

/// Read-only hint of which status effects are currently active (for rendering).
#[derive(Debug, Clone, Default, Serialize)]
#[cfg_attr(feature = "web", derive(tsify::Tsify))]
pub struct StatusHint {
    pub slow: bool,
    pub poison: bool,
    pub regen: bool,
}

impl Enemy {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        bounds: Rect,
        speed: f32,
        max_health: i32,
        score: i32,
        tower_damage: i32,
        tower_damage_interval: f32,
        path_index: usize,
        kind: EnemyKind,
    ) -> Self {
        Self {
            bounds,
            speed,
            health: max_health,
            score,
            max_health,
            tower_damage,
            tower_damage_interval,
            path_index,
            path_distance: 0.0,
            portal_cooldown_timer: CountdownTimer::new(0.0, false),
            status_effects: vec![],
            kind,
        }
    }

    pub fn position(&self) -> Vec2 {
        self.bounds.center
    }

    pub fn increase_health(&mut self, amount: i32) {
        self.health = (self.health + amount).min(self.max_health);
    }

    /// Reduce incoming bullet damage (e.g. shield).
    pub fn mitigate(&self, damage: i32) -> i32 {
        match self.kind {
            EnemyKind::Boss { shield, .. } => (damage - shield).max(0),
            _ => damage.max(0),
        }
    }

    pub fn decrease_health(&mut self, amount: i32) {
        self.health -= self.mitigate(amount);
    }

    pub fn is_destroyed(&self) -> bool {
        self.health <= 0
    }

    pub fn path<'a>(&self, paths: &'a [Path]) -> &'a Path {
        &paths[self.path_index]
    }

    /// Recompute bounds center from the current path distance.
    pub fn sync_position(&mut self, paths: &[Path]) {
        self.bounds.center = self.path(paths).position_at(self.path_distance);
    }

    /// True when the portal-teleport cooldown has elapsed (ready to teleport).
    pub fn portal_ready(&self) -> bool {
        self.portal_cooldown_timer.is_finished()
    }

    /// Arm the portal cooldown for `duration` seconds after a teleport. The cooldown
    /// mechanism prevents an enemy from moving back and forth between a portal pair.
    pub fn start_portal_cooldown(&mut self, duration: f32) {
        self.portal_cooldown_timer = CountdownTimer::new(duration, true);
    }

    /// Predicted position `time_ahead` seconds from now, following the path at
    /// current speed. Used by towers to lead moving targets.
    pub fn predicted_position(&self, time_ahead: f32, paths: &[Path]) -> Vec2 {
        self.path(paths)
            .position_at(self.path_distance + self.speed * time_ahead)
    }

    /// Resistance to slow effects (>=1.0; higher = less slowed).
    pub fn slow_resist_factor(&self) -> f32 {
        match self.kind {
            EnemyKind::Resistant { slow_resist, .. } => slow_resist,
            _ => 1.0,
        }
    }

    /// Multiplier applied to splash damage taken (<=1.0).
    pub fn splash_damage_factor(&self) -> f32 {
        match self.kind {
            EnemyKind::Resistant { splash_resist, .. } => splash_resist,
            _ => 1.0,
        }
    }

    pub fn on_death(
        &self,
        mut spawn_enemy: impl FnMut(Enemy),
        paths: &[Path],
        rng: &mut impl RngExt,
    ) {
        if let EnemyKind::Splitter { child_spec } = &self.kind {
            for _ in 0..child_spec.count {
                let mut child = Enemy::new(
                    Rect::new(self.bounds.center, child_spec.width, child_spec.height),
                    child_spec.speed,
                    child_spec.max_health,
                    child_spec.score,
                    child_spec.tower_damage,
                    child_spec.tower_damage_interval,
                    self.path_index,
                    EnemyKind::Normal,
                );

                let mut distance = self.path_distance;
                if child_spec.perturbation > 0.0 {
                    let offset =
                        rng.random_range(-child_spec.perturbation..=child_spec.perturbation);
                    // do not reach the exit
                    distance = (distance + offset)
                        .min(self.path(paths).total_length() - 0.01)
                        .max(0.0);
                }
                child.path_distance = distance;
                child.sync_position(paths);

                spawn_enemy(child);
            }
        }
    }

    pub fn apply_status_effect(&mut self, effect: StatusEffect) {
        effect.on_apply(self);
        self.status_effects.push(effect);
    }

    pub fn status_hint(&self) -> StatusHint {
        let mut hint = StatusHint::default();
        for effect in &self.status_effects {
            if !effect.expired() {
                match effect.kind {
                    StatusEffectKind::Slow { .. } => hint.slow = true,
                    StatusEffectKind::Poison { .. } => hint.poison = true,
                    StatusEffectKind::Regen { .. } => hint.regen = true,
                }
            }
        }
        hint
    }

    pub fn update(&mut self, dt: f32, paths: &[Path], map: &Map) {
        // Compute the effective speed multiplier from all active status effects and terrain.
        let slow_mult = self.status_effects.iter().fold(1.0, |acc, effect| {
            acc * if !effect.expired() {
                effect.speed_multiplier()
            } else {
                1.0
            }
        });

        // Update velocity factor and position based on terrain and status effects.
        let terrain_speed_factor = map
            .tile_at(self.position())
            .map_or(1.0, |tile| tile.enemy_speed_factor);
        let velocity_factor =
            terrain_speed_factor * (slow_mult * self.slow_resist_factor()).min(1.0);
        self.path_distance += self.speed * velocity_factor * dt;
        self.sync_position(paths);

        // Update all status effects and remove expired ones.
        let mut effects = std::mem::take(&mut self.status_effects);
        effects
            .iter_mut()
            .for_each(|effect| effect.update(self, dt));
        effects.retain(|effect| !effect.expired());

        // Update the portal cooldown timer.
        self.portal_cooldown_timer.update(dt);

        // Update boss health regeneration timer.
        if let EnemyKind::Boss {
            regen_amount,
            regen_duration,
            regen_interval,
            regen_cooldown_timer,
            ..
        } = &mut self.kind
            && *regen_interval > 0.0
            && regen_cooldown_timer.update(dt)
        {
            effects.push(StatusEffect::new(
                *regen_duration,
                StatusEffectKind::Regen {
                    regen_amount: *regen_amount,
                    accumulator: 0.0,
                },
            ));
            regen_cooldown_timer.reset();
        }

        self.status_effects = effects;
    }
}
